apps/arweave/src/ar_replica_2_9.erl

```diff
@@ -0,0 +1,473 @@
+    %% After the strict data split threshold, all end offsets are padded to a multiple of
+    %% ?DATA_CHUNK_SIZE (i.e. 262144).
+    %% --------------------------------------------------------------------------
+    
+    %% Sector start
+    assert_slice_index(1, [
+        3*262144+1, 4*262144-1, 4*262144
+    ]),
+    assert_slice_index(1, [
+        4*262144+1, 5*262144-1, 5*262144
+    ]),
+    assert_slice_index(1, [
+        5*262144+1 , 6*262144-1, 6*262144
+    ]),
+
+    %% Sector start
+    assert_slice_index(2, [
+        6*262144+1, 7*262144-1, 7*262144
+    ]),
+    assert_slice_index(2, [
+        7*262144+1, 8*262144-1, 8*262144
+    ]),
+    assert_slice_index(2, [
+        8*262144+1, 9*262144-1, 9*262144
+    ]),
+
+    %% Partition start
+    %% Sector start
+    assert_slice_index(0, [
+        9*262144+1, 10*262144-1, 10*262144
+    ]),
+    assert_slice_index(0, [
+        10*262144+1, 11*262144-1, 11*262144
+    ]),
+    assert_slice_index(0, [
+        11*262144+1, 12*262144-1, 12*262144
+    ]),
+
+    %% Sector start
+    assert_slice_index(1, [
+        12*262144+1, 13*262144-1, 13*262144
+    ]),
+
+    ?assertEqual(SubChunkCount - 1,
+            get_slice_index(get_entropy_partition_size())),
+    ?assertEqual(0,
+            get_slice_index(get_entropy_partition_size() + 1)),
+
+    ok.
+
+assert_slice_index(_ExpectedIndex, []) ->
+    ok; 
+assert_slice_index(ExpectedIndex, [AbsoluteChunkByteOffset | Rest]) ->
+    ?assertEqual(
+        ExpectedIndex, get_slice_index(AbsoluteChunkByteOffset),
+        lists:flatten(io_lib:format("get_slice_index(~p)", 
+            [AbsoluteChunkByteOffset]))
+    ),
+    assert_slice_index(ExpectedIndex, Rest).
+
+
+% @doc Walk through every sub-chunk of each chunk and verify its entropy index and
+% entropy sub-chunk index.
+entropy_index_walk_test() ->
+    ?assertEqual(3 * 8192, ?REPLICA_2_9_ENTROPY_SIZE),
+    ?assertEqual(3 * 262144, ?STRICT_DATA_SPLIT_THRESHOLD),
+    ?assertEqual(3 * 262144, get_sector_size()),
+    ?assertEqual(96, get_entropy_count()),
+    ?assertEqual(9 * 262144, get_entropy_partition_size()),
+    
+
+    %% assert_entropy_index takes a list of chunk end offsets and verifies the entropy
+    %% index for each sub-chunk in the chunk. The first argument is the expected entropy
+    %% index for the first sub-chunk in the chunk, for each subsequent sub-chunk the
+    %% expected index is incremented by 1.
+    %% 
+    %% The sector size determines the number of entropy indices. During tests the sector
+    %% size is 3*262144, so the total number of entropy indices is 3*262144 / 8192 = 96 (one
+    %% for each sub-chunk in each sector).
+    
+    %% In tests the strict data split threshold is 262144 * 3, before that offset chunks
+    %% were not padded. So each provided end offset is taken as is. After the threshold each
+    %% offset is padded to a multiple of ?DATA_CHUNK_SIZE (i.e. 262144).
```

Comment:
After the threshold is padded to threshold + a multiple of ?DATA_CHUNK_SIZE, to be precise, before determining the entropy bucket (also involves padding.)

Reply:
Updated! https://github.com/ArweaveTeam/arweave/pull/665/commits/eeff94bd8ea1c2d02d219610dee41970ab2d3f54